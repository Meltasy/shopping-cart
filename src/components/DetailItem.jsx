import { useEffect, useRef } from 'react'

const DetailItem = ({ showDetail, itemId, title, price, imageUrl, description, children }) => {
  const ref = useRef()

  useEffect(() => {
    if (!showDetail) { return }
    const dialog = ref.current
    dialog.showModal()
    return () => {
      dialog.close()
    }
  }, [showDetail])

  return (
    <>
      <dialog ref={ref}>
        <div key={itemId} className='itemDetail'>
          <h2>{title} â‚¬{price}</h2>
          <img src={imageUrl} alt=' ' />
          <p>{description}</p>
          {children}
        </div>
      </dialog>
    </>
  )
}

export default DetailItem
