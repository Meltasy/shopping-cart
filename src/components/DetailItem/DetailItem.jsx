import { useEffect, useRef } from 'react'
import styled from 'styled-components'

const StyledDialog = styled.dialog`
  padding: 0;
  border: none;
  border-radius: 10px;
  &::backdrop {
    background-color: var(--primary-color-light);
    opacity: 0.75;
  }
`

const Wrapper = styled.div`
  color: var(--text-color);
  background-color: var(--secondary-color-light);
  max-width: 500px;
  height: auto;
  padding: 1rem;
  display: flex;
  flex-direction: column;
  justify-items: center;
  align-items: center;
  gap: 0.6rem;
  @media (max-width: 480px) {
    max-width: min(90vw, 250px);
    gap: 0.3rem;
  }
`

const ExtraWrapper = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  gap: 1rem;
`

const TextWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-start;
  & h2 {
    margin: 0.5rem 0;
  }
`

const StyledImage = styled.img`
  max-width: 400px;
  max-height: 400px;
  object-fit: cover;
  border-radius: 10px;
  @media (max-width: 480px) {
    max-width: min(90vw, 200px);
    max-height: min(90vw, 200px);
  }
`

const DetailItem = ({ showDetail, itemId, title, price, imageUrl, description, children }) => {
  const ref = useRef()

  useEffect(() => {
    if (!showDetail || !ref.current) return
    const dialog = ref.current
    dialog.showModal()
    return () => {
      dialog.close()
    }
  }, [showDetail])

  return (
    <>
      <StyledDialog ref={ref}>
        <Wrapper key={itemId}>
          <ExtraWrapper>
            <TextWrapper>
              <h2>{title}</h2>
              <h2>â‚¬{price ? price.toFixed(2) : '0.00'}</h2>
            </TextWrapper>
            <div>{children}</div>
          </ExtraWrapper>
          <StyledImage src={imageUrl} alt=' ' />
          <p>{description}</p>
        </Wrapper>
      </StyledDialog>
    </>
  )
}

export default DetailItem
