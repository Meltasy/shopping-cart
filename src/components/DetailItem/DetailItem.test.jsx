import { vi, describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react'
import DetailItem from './DetailItem';

describe('Shop item detail component', () => {
  it('renders shop item detail dialog', () => {
    const { container } = render(<DetailItem />)
    expect(container).toMatchSnapshot()
  })

  it('does not render dialog when showDetail is false', () => {
    HTMLDialogElement.prototype.showModal = vi.fn()
    HTMLDialogElement.prototype.close = vi.fn()
    render(<DetailItem showDetail={false} />)
    const dialog = screen.getByRole('dialog', { hidden: true })
    expect(dialog.showModal).not.toHaveBeenCalled()
  })

  it('renders dialog when showDetail is true', () => {
    HTMLDialogElement.prototype.showModal = vi.fn()
    HTMLDialogElement.prototype.close = vi.fn()
    render(<DetailItem showDetail={true} title='Fake product' price={35.50} imageUrl='fakeImage.jpg' description='This is a fake product.' />)
    screen.debug()
    const dialog = screen.getByRole('dialog', { hidden: true })
    expect(dialog.showModal).toHaveBeenCalled()
  })

  it('renders correct item details when dialog is open', () => {
    HTMLDialogElement.prototype.showModal = vi.fn()
    HTMLDialogElement.prototype.close = vi.fn()
    render(<DetailItem showDetail={true} title='Fake product' price={35.50} imageUrl='fakeImage.jpg' description='This is a fake product.' />)
    expect(screen.getByRole('heading', { hidden: true, level: 2 })).toHaveTextContent('Fake product â‚¬35.50')
    expect(screen.getByRole('img', { hidden: true })).toHaveAttribute('src', 'fakeImage.jpg')
    expect(screen.getByText('This is a fake product.')).toBeInTheDocument()
  })
})
