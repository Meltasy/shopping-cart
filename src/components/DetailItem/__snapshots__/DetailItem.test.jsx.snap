// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop item detail component > renders shop item detail dialog 1`] = `
<div>
  <dialog
    class="sc-FEMpB hjZXnf"
  >
    <div
      class="sc-gjcoXW hKdEaB"
    >
      <div
        class="sc-dVBluf hrHbBw"
      >
        <div
          class="sc-fLDLck lhfGYe"
        >
          <h2 />
          <h2>
            â‚¬
            0.00
          </h2>
        </div>
        <div />
      </div>
      <img
        alt=" "
        class="sc-dycHyt hSmkZk"
      />
      <p />
    </div>
  </dialog>
</div>
`;
